### PR TITLE
Snapshot and Clone support

### PR DESCRIPTION
**Pull request by imran-ansari**
@wdurairaj, @nilangekarss  Please review the code

### CODE REVIEW

File: test/fake_3par_data.py [hunk 1]
```diff
@@ -0,0 +1,376 @@
...
+                                 '21810002ac00383d'),
+                   'linkState': 4}
+volume = {'name': VOLUME_NAME,
+          'id': VOLUME_ID,
+          'display_name': 'Foo Volume',
+          'size': 2,
+          'host': FAKE_CINDER_HOST,
+          'provisioning': THIN,
+          'flash_cache': None,
+          'volume_type': None,
```
Comment (nilangekarss) [inline test/fake_3par_data.py]: Do we need volume_type and volume_type_id

File: test/fake_3par_data.py [hunk 2]
```diff
@@ -0,0 +1,376 @@
...
+SNAPSHOT_3PAR_NAME = 'oss-L4I73ONuTci9Fd4ceij-MQ'
+RCG_3PAR_NAME = 'rcg-0DM4qZEVSKON-DXN-N'
+GROUP_ID = '6044fedf-c889-4752-900f-2039d247a5df'
+CONSIS_GROUP_NAME = 'vvs-YET.38iJR1KQDyA50kel3w'
+SRC_CONSIS_GROUP_ID = '7d7dfa02-ac6e-48cb-96af-8a0cd3008d47'
+SRC_CONSIS_GROUP_NAME = 'vvs-fX36AqxuSMuWr4oM0wCNRw'
+CGSNAPSHOT_ID = 'e91c5ed5-daee-4e84-8724-1c9e31e7a1f2'
+CGSNAPSHOT_BASE_NAME = 'oss-6Rxe1druToSHJByeMeeh8g'
+CLIENT_ID = "12345"
+REPLICATION_CLIENT_ID = "54321"
```
Comment (nilangekarss) [inline test/fake_3par_data.py]: Do we need replication stuff (declarations) in this class?

File: test/fake_3par_data.py [hunk 3]
```diff
@@ -0,0 +1,376 @@
...
+REPLICATION_BACKEND_ID = 'target'
+# fake host on the 3par
+FAKE_HOST = 'fakehost'
+FAKE_CINDER_HOST = 'fakehost@foo#' + HPE3PAR_CPG
+USER_ID = '2689d9a913974c008b1d859013f23607'
+PROJECT_ID = 'fac88235b9d64685a3530f73e490348f'
+VOLUME_ID_SNAP = '761fc5e5-5191-4ec7-aeba-33e36de44156'
+FAKE_DESC = 'test description name'
+FAKE_FC_PORTS = [{'portPos': {'node': 7, 'slot': 1, 'cardPort': 1},
```
Comment (nilangekarss) [inline test/fake_3par_data.py]: Do we need to define this in setup_mock.py as well?
I see duplicate definitions do we really need these duplicates?

File: test/fake_3par_data.py [hunk 4]
```diff
@@ -0,0 +1,376 @@
...
+
+volume_with_snapshots = {
+    'name': VOLUME_NAME,
+    'id': VOLUME_ID,
+    'display_name': 'Foo Volume',
+    'size': 2,
+    'host': FAKE_CINDER_HOST,
+    'provisioning': THIN,
+    'flash_cache': None,
+    'volume_type': None,
```
Comment (nilangekarss) [inline test/fake_3par_data.py]: Do we need volume_type and volume_type_id here?

File: test/fake_3par_data.py [hunk 5]
```diff
@@ -0,0 +1,376 @@
...
+    'volume_type_id': None,
+    'snapshots': [snapshot1, snapshot2, snapshot3]}
+
+volume_replicated = {'name': VOLUME_NAME,
```
Comment (nilangekarss) [inline test/fake_3par_data.py]: can you remove this dictionary if its not needed

File: test/fake_3par_data.py [hunk 6]
```diff
@@ -0,0 +1,376 @@
...
+                               'qos:maxIOPS': '1000',
+                               'qos:maxBWS': '50',
+                               'qos:minIOPS': '100',
+                               'qos:minBWS': '25',
+                               'qos:latency': '25',
+                               'qos:priority': 'low'},
+               'deleted_at': None,
+               'id': 'gold'}
+
+volume_type_replicated = {'name': 'replicated',
```
Comment (nilangekarss) [inline test/fake_3par_data.py]: Can this be removed so that we have only needed definitions and declarations

File: test/fake_3par_data.py (hunk 1 above)
Comment (imran-ansari) [reply test/fake_3par_data.py]: We don't actually. Will remove it.

File: test/fake_3par_data.py (hunk 2 above)
Comment (imran-ansari) [reply test/fake_3par_data.py]: We don't - will remove it.

File: hpedockerplugin/hpe/hpe_3par_common.py [hunk 7]
```diff
@@ -633,6 +657,9 @@ def create_volume(self, volume):
             extras = {'comment': json.dumps(comments),
                       'tpvv': tpvv, }
 
+            if len(self.config.hpe3par_snapcpg):
+                extras['snapCPG'] = self.config.hpe3par_snapcpg[0]
+
```
Comment (hpe-storage) [inline hpedockerplugin/hpe/hpe_3par_common.py]: if the config param for snapCPG is not given (in hpe.conf) can we default the snapcpg to the actual cpg where the volume will be created ?

File: hpedockerplugin/hpe/hpe_3par_common.py [hunk 8]
```diff
@@ -938,3 +968,202 @@ def _get_prioritized_host_on_3par(self, host, hosts, hostname):
...
+                    vol_chap_enabled = self.client.getVolumeMetaData(
+                        src_3par_vol_name, 'HPQ-docker-CHAP-name')['value']
+                except hpeexceptions.HTTPNotFound:
+                    LOG.debug("CHAP is not enabled on volume %(vol)s ",
+                              {'vol': src_vref['id']})
+                    vol_chap_enabled = False
+
+            # # Check whether a process is a backup
+            # if str(src_vref['status']) == 'backing-up':
+            #     back_up_process = True
```
Comment (hpe-storage) [inline hpedockerplugin/hpe/hpe_3par_common.py]: These statements seems to come from cinder driver. Please remove them

File: hpedockerplugin/hpe/hpe_3par_common.py [hunk 9]
```diff
@@ -938,3 +968,202 @@ def _get_prioritized_host_on_3par(self, host, hosts, hostname):
...
+
+            # if the sizes of the 2 volumes are the same and except backup
+            # process for ISCSI volume with chap enabled on it.
+            # we can do an online copy, which is a background process
+            # on the 3PAR that makes the volume instantly available.
+            # We can't resize a volume, while it's being copied.
+            if dst_volume['size'] == src_vref['size'] and not (
+               #back_up_process and
```
Comment (hpe-storage) [inline hpedockerplugin/hpe/hpe_3par_common.py]: remove this commented line

File: hpedockerplugin/hpe_storage_api.py [hunk 10]
```diff
@@ -221,6 +246,99 @@ def volumedriver_remove(self, name):
...
+                        return response
+                else:
+                    msg = (_LE('snapshot %s does not exist!' % snapname))
+                    LOG.error(msg)
+                    # raise exception.HPEPluginRemoveException(reason=msg)
+                    response = json.dumps({u"Err": msg})
+                    return response
+
+        except Exception:
+                LOG.debug('volume: %(name)s is locked',
```
Comment (hpe-storage) [inline hpedockerplugin/hpe_storage_api.py]: change this to LOG.error()

File: hpedockerplugin/hpe_storage_api.py [hunk 11]
```diff
@@ -221,6 +246,99 @@ def volumedriver_remove(self, name):
...
+                response = json.dumps({u"Err": ''})
+                return response
+        finally:
+            # Expand lock code inline as function based lock causes
+            # unexpected behavior
+            try:
+                self._etcd.try_unlock_volname(volname)
+            except Exception as ex:
+
+                LOG.debug('volume: %(name)s Unlock Volume Failed',
```
Comment (hpe-storage) [inline hpedockerplugin/hpe_storage_api.py]: change this LOG.error

File: test/fake_3par_data.py (hunk 3 above)
Comment (imran-ansari) [reply test/fake_3par_data.py]: fake_3par_data.py was introduced later. Hence you are seeing residue of earlier implementation. Will remove those.

File: test/fake_3par_data.py (hunk 5 above)
Comment (imran-ansari) [reply test/fake_3par_data.py]: Sure

File: test/fake_3par_data.py (hunk 6 above)
Comment (imran-ansari) [reply test/fake_3par_data.py]: Sure

File: hpedockerplugin/hpe/hpe_3par_common.py (hunk 7 above)
Comment (imran-ansari) [reply hpedockerplugin/hpe/hpe_3par_common.py]: Okay.

File: hpedockerplugin/hpe/hpe_3par_common.py (hunk 8 above)
Comment (imran-ansari) [reply hpedockerplugin/hpe/hpe_3par_common.py]: Removed.

File: hpedockerplugin/hpe/hpe_3par_common.py (hunk 9 above)
Comment (imran-ansari) [reply hpedockerplugin/hpe/hpe_3par_common.py]: Okay

File: hpedockerplugin/hpe_storage_api.py (hunk 10 above)
Comment (imran-ansari) [reply hpedockerplugin/hpe_storage_api.py]: Done

File: hpedockerplugin/hpe_storage_api.py (hunk 11 above)
Comment (imran-ansari) [reply hpedockerplugin/hpe_storage_api.py]: Done

**Pull request merged**